target/riscv/cpu_helper.c

```diff
@@ -743,7 +745,18 @@ static int get_physical_address(CPURISCVState *env, hwaddr *physical,
                     }
                 }
             }
-            if ((pte & PTE_CW) == 0) {
+
+            /*
+             * Flag this TLB entry as requiring a trap on cap-store if either
+             * CD or CW is clear.  This isn't great, since the check for
+             * PAGE_SC_TRAP is off the TCG-native path and so will actaully
+             * trigger a trap when all we really want is a new lookup, but
+             * it beats failing to update CD if the first access to a page
+             * isn't a cap store.
```

Comment:
I mean, it's an improvement, because we have software emulation of CD for our various soft cores, but it's not great, and it's supposed to be that if hardware dirty tracking is supported then so is capability dirty tracking (though at a glance it doesn't look like we actually specify that). Can we not just fix this properly so PAGE_SC_TRAP leads to a TLB miss like a missing PAGE_WRITE (or invert it so a missing PAGE_CAP_WRITE leads to a TLB miss)? The LC handling is awkward due to the additional state, but SC doesn't have to be so messy.

Reply:
After 818c573e88d64b042225ceb2515e04888b7b8eff, this does the right thing.  I've updated the comment to match.